src/cas.rs

```diff
@@ -1,108 +1,93 @@
-        if let Some(annotations) = self.annotations.get(name) {
-            for annot in annotations {
-                let covered_text = self
-                    .get_covered_text_safe(annot.begin as usize, annot.end as usize)
-                    .unwrap();
-                println!(
-                    "ANNOT {} BEGIN {} END {}\nCOVERED TEXT IS: {}",
-                    name, annot.begin, annot.end, covered_text
-                );
-            }
-        } else {
-            return Err(PipelineError::AnnotationMissing);
+
+    pub fn print_annotations(&self, name: &str) -> Result<()> {
+        let annotations = self
+            .annotations
+            .get(name)
+            .ok_or(PipelineError::AnnotationMissing)?;
+
+        for annot in annotations {
+            let covered_text = &self.text[annot.begin..annot.end];
+            println!(
+                "ANNOT {} BEGIN {} END {}\nCOVERED TEXT IS: {}",
+                name, annot.begin, annot.end, covered_text
+            );
         }
+
         Ok(())
     }
+
     pub fn get_covered_annotations(
         &self,
         annot_name: &str,
-    ) -> Result<HashMap<String, Vec<&Annotation>>, PipelineError> {
+    ) -> Result<HashMap<String, Vec<&Annotation>>> {
         let mut covered_annotations: HashMap<String, Vec<&Annotation>> = HashMap::new();
-        if let Some(annot_cover) = self.annotations.get(annot_name) {
-            for annot in annot_cover {
-                let begin = annot.begin;
-                let end = annot.end;
-                for (name, annotations) in &self.annotations {
-                    if name != annot_name {
-                        for a in annotations {
-                            if a.begin >= begin && a.end <= end {
-                                covered_annotations
-                                    .entry(name.to_string())
-                                    .or_insert_with(|| Vec::new())
-                                    .push(a);
-                            }
-                        }
-                    }
+        let annot_cover = self
+            .annotations
+            .get(annot_name)
+            .ok_or(PipelineError::AnnotationMissing)?;
+
+        for annot in annot_cover {
+            for (name, annotations) in &self.annotations {
+                if name != annot_name {
+                    covered_annotations
+                        .entry(name.clone())
+                        .or_default()
+                        .extend(annotations.iter().filter(|a| annot.covers(a)));
                 }
             }
-        } else {
-            return Err(PipelineError::AnnotationMissing);
         }
 
         Ok(covered_annotations)
     }
+
     pub fn get_covered_annotations_by(
         &self,
         annot_name: &str,
         annot_covered: &str,
-    ) -> Result<HashMap<String, Vec<&Annotation>>, PipelineError> {
+    ) -> Result<HashMap<String, Vec<&Annotation>>> {
```

Comment:
Does this need to be a hashmap? If I'm looking at it right, the map only ever gets a key for `annot_covered`, so this function could just return a `Vec<&Annotation>` instead.